platform/default/headless_backend_osmesa.cpp

```diff
@@ -1,46 +1,64 @@
 #include <mbgl/platform/default/headless_backend.hpp>
 #include <mbgl/platform/log.hpp>
 
+#include <GL/osmesa.h>
+
+#include <cassert>
+
 namespace mbgl {
 
+struct HeadlessBackend::Impl {
+    OSMesaContext glContext = nullptr;
+    GLubyte fakeBuffer = 0;
+};
+
+void HeadlessBackend::ImplDeleter::operator()(Impl* impl_) const {
+    OSMesaDestroyContext(impl_->glContext);
+    delete impl_;
+}
```

Comment:
Instead of using custom deleters, we could also use subclasses and put them into the `unique_ptr`.